apollo-federation/src/query_plan/optimize.rs

```diff
@@ -0,0 +1,755 @@
+    // and is likely the most common case of this kind of inclusion. But in theory, we would
+    // have `fragment F1 on X { a b c }` and `fragment F2 on X { b c }`, in which case `F2` is
+    // still included in `F1`, but we'd have to work harder to figure this out and it's unclear
+    // it's a good tradeoff. And while you could argue that it's on the user to define its
+    // fragments a bit more optimally, it's actually a tad more complex because we're looking
+    // at fragments in a particular context/parent type. Consider an interface `I` and:
+    // ```graphql
+    //   fragment F3 on I {
+    //     ... on X {
+    //       a
+    //     }
+    //     ... on Y {
+    //       b
+    //       c
+    //     }
+    //   }
+    //
+    //   fragment F4 on I {
+    //     ... on Y {
+    //       c
+    //     }
+    //     ... on Z {
+    //       d
+    //     }
+    //   }
+    // ```
+    // In that case, neither fragment include the other per-se. But what if we have
+    // sub-selection `{ b c }` but where parent type is `Y`. In that case, both `F3` and `F4`
+    // applies, and in that particular context, `F3` is fully included in `F4`. Long story
+    // short, we'll currently return `{ ...F3 ...F4 }` in that case, but it would be
+    // technically better to return only `F4`. However, this feels niche, and it might be
+    // costly to verify such inclusions, so not doing it for now.
+    fn reduce_applicable_fragments(
+        applicable_fragments: &mut Vec<(Node<Fragment>, FragmentRestrictionAtType)>,
+    ) {
+        // Note: It's not possible for two fragments to include each other. So, we don't need to
+        //       worry about inclusion cycles.
+        let included_fragments: HashSet<Name> = applicable_fragments
+            .iter()
+            .filter(|(fragment, _)| {
+                applicable_fragments
+                    .iter()
+                    .any(|(other_fragment, _)| other_fragment.includes(&fragment.name))
+            })
+            .map(|(fragment, _)| fragment.name.clone())
+            .collect();
+
+        applicable_fragments.retain(|(fragment, _)| !included_fragments.contains(&fragment.name));
+    }
+
+    /// Try to optimize the selection set by re-using existing fragments.
+    /// Returns either
+    /// - a new selection set partially optimized by re-using given `fragments`, or
+    /// - a single fragment that covers the full selection set.
+    // PORT_NOTE: Moved from `Selection` class in JS code to SelectionSet struct in Rust.
+    // PORT_NOTE: `parent_type` argument seems always to be the same as `self.type_position`.
+    fn try_optimize_with_fragments(
+        &self,
+        parent_type: &CompositeTypeDefinitionPosition,
+        fragments: &NamedFragments,
+        validator: &mut FieldsConflictMultiBranchValidator,
+        can_use_full_matching_fragment: impl Fn(&Fragment) -> bool,
```

Comment:
let's just have this as a function on `Fragment` directly.

Reply:
This has been improved in FED-191. Please bear with me on this for now.